### PR TITLE
Document security scope for extensions

### PR DESCRIPTION
**Pull request by peternied**
### Description
While  working on extensions there are many considerations for security, this document captures what those areas are as a way to discuss design approaches and codify the current state of affairs.

To approve this document means that we have `agreement of areas and 'as-is' state of OpenSearch Plugins and Extensions.` There are more details as to the phases of this document over time within the doc itself

### Issues Resolved
* Related https://github.com/opensearch-project/opensearch-sdk/issues/24
* Related https://github.com/opensearch-project/security/issues/1895

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by peternied** [comment]
@opensearch-project/security @CEHENKLE @dblock @saratvemulapalli @dbwiddis What do you think of using a document like this for extensions security related features?

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:3]: ...expanding on them. With this increased functionality comes a larger surface area....

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:13]: I'm unclear about "is the mechanism" vs. "it's deprecated".    If it's on its way out this should be much more past-tense focused, e.g., "While the Java Security Manager has been.... " and add on "Additional measures are needed to protect system resources."

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:15]: REST should be all caps.  While isolating is a benefit it also brings with it the requirement for extensions to follow strict APIs which may enforce limits.

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:19]: Let's avoid passive voice here: The OpenSearch cluster transfers data to the extensions via https requests between the nodes on the cluster and the extensions endpoint(s).

Does this explain two way communications and how the extensions themselves communicate?  I see Extension <--->  Extension points <----> Cluster.

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:21]: Let's add the why.  "To ensure security and identity best practices are followed ..."

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:25]: Is the security plugin going to remain as-is? That's not clear here.  I think it's a good idea if so but we should articulate that.

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:27]: Plugins do store, or may store?

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:31]: This needs a lot more detail, or perhaps just a link to the security plugin docs.

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:33]: This needs an "As is" and "To Be" summary.  We are planning to do something else with User, aren't we?  You're also mixing extension and plugin here, not sure if this is documenting current vs. future.   I would like to put RBAC as a goal.

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:35]: Possibly currently.  In the future I would want permissions and roles to be identity-based using open standards for RBAC.

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:47]: Should we mention Semantic Versioning standards?

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:43]: This seems to be an implementation detail that we don't need to specify here.

**Review by dbwiddis — COMMENTED** [review]
Good first start. I like the distinction between plugin and extension up front but then I'm starting to miss the usage later on as everything seems to say "plugin". Is it documenting current state?  

It may be good to organize the sections into "current state" (plugin) and "future state" (extension + select plugins) to make clear what we are scoping for in the future.

**Comment by dbwiddis** [comment]
> @opensearch-project/security @CEHENKLE @dblock @saratvemulapalli @dbwiddis What do you think of using a document like this for extensions security related features?

I think a _template_ with a subset of important decision points similar to a document like this is a good idea once this document is ironed out.   This document can serve more as a guiding principle while extensions should complete the template with implementation specifics for that extension.

File: SECURITY.md
Comment (peternied) [reply SECURITY.md:19]: 💯 on the passive voice comment, thanks for noticing

File: SECURITY.md
Comment (peternied) [reply SECURITY.md:13]: Rewording to make this clearer - let me know what you think of the revision

File: SECURITY.md
Comment (peternied) [reply SECURITY.md]: Reworked this section, does this sound better?

File: SECURITY.md
Comment (peternied) [reply SECURITY.md]: > I would like to put RBAC as a goal.

Me too, I think there is a version of this document that includes lists of issues to be implemented/decisions to make.  Before we come to the possible implementations I'd like to lay out the surface area for prioritization.  How would you feel about holding off on individual features until we codify this framework?

File: SECURITY.md
Comment (peternied) [inline SECURITY.md:7]: @dbwiddis Following your feedback added this section to capture how to approve/reject this document.

File: SECURITY.md
Comment (peternied) [reply SECURITY.md:57]: OpenSearch hardly has requirements in this space right now, here are some areas of future recommendations:
* authenticity at installation, dev plugin vs self-signed vs from a registry
* authenticity during runtime, how OpenSearch is sure the extension endpoint didn't change, or vice-versa
* installation gates, allow some plugins to be install but not any plugin
* controlling installed extensions capabilities, restrict what a plugin has access to do

File: SECURITY.md
Comment (davidlago) [inline SECURITY.md:29]: This is only true for out-of-process extensions. How about the ones that can run in-process? I think we are still open to having those supported down the line

File: SECURITY.md
Comment (davidlago) [inline SECURITY.md:45]: I think the word proprietary might not be the one you're looking for here (see https://en.wikipedia.org/wiki/Proprietary_file_format). Perhaps "specific to OpenSearch", "non-standard" (if there is a standard for these...?

File: SECURITY.md
Comment (davidlago) [inline SECURITY.md:49]: this is another statement that might need to be revisited if we allow in-process extensions

File: SECURITY.md
Comment (davidlago) [inline SECURITY.md:59]: I'm not clear on what this means (and specifically how it relates to security)

File: SECURITY.md
Comment (davidlago) [inline SECURITY.md:67]: Is this a security aspect?

File: SECURITY.md
Comment (davidlago) [inline SECURITY.md:71]: Do we also use OpenSearch indexes to store configuration elements too?

File: SECURITY.md
Comment (peternied) [reply SECURITY.md:29]: AFAIK in process extensions are not being designed or implemented at this time.  How do you feel about not including it until this effort is more concrete?

File: SECURITY.md
Comment (peternied) [reply SECURITY.md:59]: This is relevant in that CVEs for plugins are tightly coupled with one another and OpenSearch.  Whereas extensions and OpenSearch are be able to perform minor/patch revisions for CVEs without coordination.  I've update the section to better capture this.

File: SECURITY.md
Comment (peternied) [reply SECURITY.md:71]: Good point, I know the security plugin built functionality for this pattern, but I'm not sure if other plugins do the same.  I'll follow up on this.

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:27]:
```suggestion
Plugins depend on use of the Java Security Manager (JSM) to limit interactions on the host operation system resources (cpu/disk/memory/network/...).  JSM has been deprecated, with its removal scheduled in the next release of the JVM, see [OpenSearch discussion](https://github.com/opensearch-project/OpenSearch/issues/1687). Additional measures are needed to protect system resources.
```

File: SECURITY.md
Comment (dbwiddis) [reply SECURITY.md:29]: I think defining "REST API" is too specific for the potential application. The key point to make here is that interaction is limited to a defined set of API locations.  For external extensions, they will use REST.  If REST is not needed (e.g., in-process) there still should be another method of using the same API point.

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:39]: If it's unencrypted we should explain how it is presently "secured".

File: SECURITY.md
Comment (dbwiddis) [inline SECURITY.md:47]:
```suggestion
For resource that are managed by plugins, access control is governed within individual plugin. By examining the [user](https://github.com/opensearch-project/common-utils/blob/main/src/main/java/org/opensearch/commons/authuser/User.java) object from OpenSearch's thread context, permissions are available for approval/denial. An example from anomaly detection is [checkUserPermissions](https://github.com/opensearch-project/anomaly-detection/blob/875b03c1c7596cb34d74fea285c28d949cfb0d19/src/main/java/org/opensearch/ad/util/ParseUtils.java#L568).  Uniform resource controls and models are needed to protect from misconfiguration and code defects.
```

**Comment by dbwiddis** [comment]
I've been doing some thinking about the integration of security, extensions, and versioning and think they all inter-relate a bit.  Tossing the following diagram and content which you might consider including in the scope document in some way.  Presently an outline, but each of the below bullet points should be fleshed out.  @peternied @saratvemulapalli interested in your thoughts on this.

image

## API Controls

All of the below interactions will use the API/SDK interaction points as the primary means of enabling/enforcing the points below.
 - SDK will ensure Extensions identify themselves with a unique version which is permitted, Users authenticate appropriately, and requests are rejected if they would potentially harm the cluster

### Security & Identity

 - The OpenSearch cluster is protected from improper use of its resources
 - Only authorized users have accesess to data that they need in their current role (RBAC)

### Versioning

 - All components of the OpenSearch ecosystem (core, extensions) can be uniquely identified in a way that represents their features as well as code, dependencies, etc.

### Extensibility

 - Code external to OpenSearch can communicate with OpenSearch via a defined API/SDK

#### Security + Versioning

 - Individual versions of extensions can be declared unsecure (alternately, extensions must be "approved" to be allowed) due to vulnerabilities

#### Security + Extensibility

 - Extensions need a common framework to communicate user identities to the API, including certificate stores, identity server, etc.

#### Extensiblity + Versioning

 - Semantic Versioning and feature sets should combine to provide compatible version ranges

File: SECURITY.md
Comment (peternied) [reply SECURITY.md:39]: The host security is the only check on this data access; we could callout this overlap.  As it occurs to me, the data is not secured in a sufficient manor to justify the use of the term 'secure'.  

Services such as AWS' managed OpenSearch do have at rest encryption for data, and I wanted to make sure that readers of this doc did not conflate the two as I did when I first started writing this doc.

What do you think of leaving this text as it is?

File: SECURITY.md
Comment (peternied) [reply SECURITY.md:29]: If extensions are available in-process, it seems like we are recreating the existing plugin feature with its same architectural weaknesses.   With a solid security boundary between OpenSearch and its extensions, we can build extensibility in alignment with trustworthy functionality.  I think there is a discussion to be had for in-process extensions and I would prefer to wait to add those considerations to this document until we are committing more thought in this space, does this seem like a worthwhile trade-off in the near term?

I am only away of using REST APIs for extensions, if we want to broaden the language do you have a suggestion?

File: SECURITY.md
Comment (dbwiddis) [reply SECURITY.md]: As mentioned in earlier comment, we do need an individual feature list somewhere. We can limit this particular document to the overall framework/scope but identifying at least major categories of features is needed for any actual planning and scoping of future work

File: SECURITY.md
Comment (dbwiddis) [reply SECURITY.md:29]: Looked into this further and I'm fine with leaving this as REST API for simplicity.

**Review by saratvemulapalli — COMMENTED** [review]
Thanks for writing this up @peternied.

Couple of more thoughts/questions:
- How do we protect inter-extension communication? 
- How can we enable extensions to expose resources, perform authentication and authorization?
- I see across the docs, the assumption is that the communication based on REST API's. So far extensions are just built on transport(TCP). For security, if REST APIs are preferred we could weigh in pros and cons.
- I see this doc scope as security for plugins and extensions. What are we thinking for Security for OpenSearch? 
- When we define scope, i believe it would be the time to define tenets we would want to stick to. 
- We did close: https://github.com/opensearch-project/security/issues/1895 in favor for this PR. I dont see the tactical items on the plate to execute for supporting AD as an extension. How do you want to take on this?

File: SECURITY.md
Comment (peternied) [reply SECURITY.md]: It looks like this was resolved in a later revision, could you double check?

File: SECURITY.md
Comment (peternied) [reply SECURITY.md]: Instead of going into this detail, I've linked the blog post.  As we plan to invest more energy into areas like this we can flesh them out better.

**Comment by peternied** [comment]
Great questions @saratvemulapalli thanks for the feedback.

> How do we protect inter-extension communication?

This sounds like a future-feature request.  For the moment, proxied via OpenSearch seems like the path, but there is much potential in this space.  When we consider adding this functionality I think it warrants an in-depth discussion.  Should we file an issue on this now to track, or is there another document that captures this?  I'm leaning on having this come back up organically instead of creating an issue we forget about.

> How can we enable extensions to expose resources, perform authentication and authorization?

I'm not sure that we would implement all of these in this SDK.  For example in the identity space there is [Open ID](https://openid.net/what-is-openid/), I do not think that functionality would below in this SDK, as implementing the standard in a native plugin might be the better way.

> I see across the docs, the assumption is that the communication based on REST API's. So far extensions are just built on transport(TCP). For security, if REST APIs are preferred we could weigh in pros and cons.

If we want to be security focused with extensions, I think we need to strongly consider moving to only communicate via REST API.  While there are likely trade-offs from existing functionality/perf, there are huge benefits to quickly onboard simple extensions written in python, rust, or golang.  Additionally, by limited communicates with an ephemeral connections we can be sure that authorization is granted at the time of a given request.  It creates a well defined trust boundary.  I've created https://github.com/opensearch-project/opensearch-sdk/issues/34 for additional discussion, as well as updating this doc in the COMSEC section.

> I see this doc scope as security for plugins and extensions. What are we thinking for Security for OpenSearch?

This is a good callout, I'll start an effort to connect OpenSearch / Security as well as this doc together.

> When we define scope, i believe it would be the time to define tenets we would want to stick to.

I like this in principle, but I am struggling to see how we could roll this out.  I don't think we have existing engineering tenets in OpenSearch, this might be exactly the time to create them.  This said I think that those will be documented in the OpenSearch codebase and I'd like to see about merging this change without them, how does that occur to you?

> We did close: https://github.com/opensearch-project/security/issues/1895 in favor for this PR. I dont see the tactical items on the plate to execute for supporting AD as an extension. How do you want to take on this?

Thanks for bringing this up, after I merge this version of the document, there should be a fast follow for calling out the investigations and work that needs to be done to support Anomaly Detections feature set.  I'll reopen that issue and close it when this has been done.

**Review by dbwiddis — APPROVED** [review]

**Review by saratvemulapalli — APPROVED** [review]

**Comment by dbwiddis** [comment]
> @dbwiddis @saratvemulapalli how does this look in its current form?

I'm happy with it in it's current form.  We may tweak a few things in the future but it's a good working document at this point.

Good questions regarding REST and auth/security.  I think it's a solvable problem, and all these are implementation details rather than scope which the doc outlines.
 - We can use SSL/TLS for the REST APIs to encrypt and then use (initially) the existing security mechanisms (user/password, certificate based) while also enabling (future) identity/RBAC capabilities.
   - May consider whether to make this user configurable, e.g., force https by default/allow insecure opt-in.
 - In-process comms can still use the REST API; localhost avoids any network traffic
 - Inter-process comms we'll need to figure out how to register availability, push vs. pull, robustness to failures, etc.

File: SECURITY.md
Comment (peternied) [reply SECURITY.md]: Lets revisit this in a follow up PR to make sure we've got the detail we need

**Pull request merged**
